## Importing libraries

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

## Importing Dataframes

In [2]:
# Define path
path = r'/Users/isaacotubanjo/Documents/08:08:2023 Instacart Business Analysis'

In [3]:
# Import orders_products_with_new_variables.pkl as df_ords_prods_merge
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_with_new_variables.pkl'))

## Using the GroupBy Function

In [4]:
# Using groupby function and aggregating the mean 
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

The mean here varies less than the mean in the subset (difference between the highest and lowest means in the subset is 8.17 while the difference in the entire dataframe is 7.69. This is even reduced further when you take out the department id 21 which represents the missing category and has the highest mean). The values are also slightly reduced (for example: department id 4 with 17.81 in the full dataframe and 18.82 in the subset).

## Aggregating Data with Transform

In [6]:
# Splitting into groups and generating a new column 
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [7]:
# To confirm this was executed successfully, we can view the first 50 rows of the dataframe
pd.options.display.max_rows = None
# This will remove the view restriction on Python

In [8]:
df_ords_prods_merge.head(50)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   
20   2231262       31            17                  3                 11   
21   2580647       43             6                  4                 16   
22   2187180       43             9                  4                 12   
23   2497897       52             1                  1                  9   
24   1318871       52             2                  1                 10   
25   1261384       52             3                  1                 10   
26    580568       52             5                  2                 10   
27   2428073       52             7                  2                  7   
28    342306       52             9                  1                 11   
29   1520353       52            10                  2                 10   
30   2542086       52            11                  2                  9   
31    944694       52            12                  1                 16   
32   1268191       52            15                  2                 11   
33   1122089       52            17                  2                 10   
34   1498922       52            18                  2                 10   
35    180919       52            19                  3                  8   
36   2695875       52            21                  5                 10   
37   3205960       67             1                  3                 11   
38    970922       67             2                  1                 15   
39    596585       67             3                  1                 13   
40   2037087       67             4                  2                  9   
41   2474111       67             5                  3                 14   
42   1568481       67             6                  3                 12   
43   2048527       67             8                  2                 11   
44   2389031       67             9                  2                 14   
45   2378562       67            10                  1                 10   
46   1515698       67            11                  1                 11   
47   1928050       67            12                  2                 12   
48   1949274       67            13                  1                 10   
49    551869       67            14                  1                  9   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0    

The new column (max_order) has been created 

## Creating a loyalty flag for existing customers

In [9]:
# Creating a new column with loc
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [10]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [11]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
df_ords_prods_merge.head(5)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

  busiest_day_of_week        busiest_days busiest_period_of_day  max_order  \
0      Regularly busy  Regularly busy day        Average orders         10   
1      Regularly busy     Least busy days        Average orders         10   
2      Regularly busy     Least busy days           Most orders         10   
3          Least busy     Least busy days        Average orders         10   
4          Least busy     Least busy days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [13]:
# Confirming the sum of all 3 new categories of customers created in the loyalty_flag column
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [14]:
df_ords_prods_merge.shape

(32404859, 19)

The total number of customers based on loyalty is the same as the previous total so the code has been successfully executed.

## Determining the prices of products purchased by each customer loyalty group

In [15]:
# Grouping by prices
df_ords_prods_merge.groupby('prices')

In [16]:
# Split the data into groups based on loyalty and then calculate the mean for each "prices" column
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

Here we can see that the new customers tend to buy the more expensive products on average. There is not much difference and as we already know that the loyal customers tend to buy more, I believe they are still the most important group.

## Determining the average prices spent by customers

In [17]:
# Splitting into groups and generating a new column for average_price
df_ords_prods_merge['avg_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)
# This will create a new column with the average price for each user_id

In [18]:
df_ords_prods_merge.head(50)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   
20   2231262       31            17                  3                 11   
21   2580647       43             6                  4                 16   
22   2187180       43             9                  4                 12   
23   2497897       52             1                  1                  9   
24   1318871       52             2                  1                 10   
25   1261384       52             3                  1                 10   
26    580568       52             5                  2                 10   
27   2428073       52             7                  2                  7   
28    342306       52             9                  1                 11   
29   1520353       52            10                  2                 10   
30   2542086       52            11                  2                  9   
31    944694       52            12                  1                 16   
32   1268191       52            15                  2                 11   
33   1122089       52            17                  2                 10   
34   1498922       52            18                  2                 10   
35    180919       52            19                  3                  8   
36   2695875       52            21                  5                 10   
37   3205960       67             1                  3                 11   
38    970922       67             2                  1                 15   
39    596585       67             3                  1                 13   
40   2037087       67             4                  2                  9   
41   2474111       67             5                  3                 14   
42   1568481       67             6                  3                 12   
43   2048527       67             8                  2                 11   
44   2389031       67             9                  2                 14   
45   2378562       67            10                  1                 10   
46   1515698       67            11                  1                 11   
47   1928050       67            12                  2                 12   
48   1949274       67            13                  1                 10   
49    551869       67            14                  1                  9   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0    

## Creating a spending flag for customers based on average prices spent

In [19]:
# Creating a new column with loc
df_ords_prods_merge.loc[df_ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [20]:
df_ords_prods_merge.loc[df_ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'

In [22]:
# Checking the sum of the values under spending_flag to confirm accuracy 
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

There are 31770614 low spenders and 634245 high spenders. The high spenders are the majority with 98% of the total number of customers.

In [24]:
df_ords_prods_merge.shape

(32404859, 21)

The total number of customers based on spending is the same as the previous total so the code has been successfully executed. Also, we have a lot more low spenders than high spenders based on average prices.

## Determining the frequency of customers' shopping based on days since last order

In [25]:
# Splitting into groups and generating a new column for median days
df_ords_prods_merge['median_days'] = df_ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)
# This will create a new column with the median days for each user_id

In [26]:
df_ords_prods_merge.head(50)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   
20   2231262       31            17                  3                 11   
21   2580647       43             6                  4                 16   
22   2187180       43             9                  4                 12   
23   2497897       52             1                  1                  9   
24   1318871       52             2                  1                 10   
25   1261384       52             3                  1                 10   
26    580568       52             5                  2                 10   
27   2428073       52             7                  2                  7   
28    342306       52             9                  1                 11   
29   1520353       52            10                  2                 10   
30   2542086       52            11                  2                  9   
31    944694       52            12                  1                 16   
32   1268191       52            15                  2                 11   
33   1122089       52            17                  2                 10   
34   1498922       52            18                  2                 10   
35    180919       52            19                  3                  8   
36   2695875       52            21                  5                 10   
37   3205960       67             1                  3                 11   
38    970922       67             2                  1                 15   
39    596585       67             3                  1                 13   
40   2037087       67             4                  2                  9   
41   2474111       67             5                  3                 14   
42   1568481       67             6                  3                 12   
43   2048527       67             8                  2                 11   
44   2389031       67             9                  2                 14   
45   2378562       67            10                  1                 10   
46   1515698       67            11                  1                 11   
47   1928050       67            12                  2                 12   
48   1949274       67            13                  1                 10   
49    551869       67            14                  1                  9   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0    

## Creating an order frequency flag based on median days since last order

In [27]:
# Creating a new column with loc
df_ords_prods_merge.loc[df_ords_prods_merge['median_days'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [29]:
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days'] <= 20) & (df_ords_prods_merge['median_days'] > 10), 'order_frequency_flag'] = 'Regular customer'

In [30]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [31]:
# Checking the sum of the values under order_frequency_flag to confirm accuracy 
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

We have a majority of frequent customers based on median days but we also have 5 under NaN (unknown values)

In [33]:
# To find out what category the NaN values fall under
print(df_ords_prods_merge[df_ords_prods_merge.order_frequency_flag.isnull()])
# This will return all rows where the order_frequency_flag is null

          order_id  user_id  order_number  order_day_of_week  \
13645692    895835   159838             1                  0   
17251990    895835   159838             1                  0   
17622767    895835   159838             1                  0   
24138593    895835   159838             1                  0   
25880002    895835   159838             1                  0   

          order_hour_of_day  days_since_last_order  product_id  \
13645692                 17                    NaN       10749   
17251990                 17                    NaN       33401   
17622767                 17                    NaN       23695   
24138593                 17                    NaN       21334   
25880002                 17                    NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0   
25880002

As confirmed earlier, there are 5 rows with null values. These values fall under just 1 user_id (159838) with one order_id (895835). This means that this customer has only had one order and therefore we can't calculate median days. Since the criteria for grouping states that grouping have to be made based on median days, I believe we can categorise this customer under 'Non-frequent' as they have never returned to shop.

In [34]:
# we can change the null values in "order_frequency_flag" column to Non-frequent customer
df_ords_prods_merge['order_frequency_flag'].fillna("Non-frequent customer", inplace = True)
# This will replace all NaN values

In [35]:
# Checking the sum of the values under order_frequency_flag to confirm accuracy 
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636442
Name: order_frequency_flag, dtype: int64

Now, there are no NaN values anymore

We have 21,559,853 frequent customers and 3,636,442 non-frequent customers. The frequent customers are the majority with 66.5% of the total customer numbers. Non-frequent customers occupy just 1.1% of the total customer numbers.

## Exporting dataframe to pickle

In [36]:
# Export data to pickle
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_aggregated.pkl'))